### PR TITLE
Feat: Refactor for handling the checker states by functions. And Fix for autofill.  

### PR DESCRIPTION
**Pull request by svenjens**
We refactored the wrapper page to use functions instead of updating the context directly. 
We can now call `isActive('conclusion')` to check if the component is active. 

Also we fixed autofill in this branch, and some minor changes.

### CODE REVIEW

File: apps/client/src/pages/WrapperPage.js [hunk 1]
```diff
@@ -30,35 +67,35 @@ const WrapperPage = ({ checker, topic }) => {
         highlightActive
       >
         <StepByStepItem
-          active={!finishedLocation}
-          checked={finishedLocation}
+          active={() => { isActive('location') || isActive('address')}}
+          checked={() => isFinished('location', 'address')}
```
Comment (afjlambert) [inline apps/client/src/pages/WrapperPage.js]: I don't think this could actually work. `isFinished` uses `Array.includes` on the first argument only. You're passing two.

File: apps/client/src/pages/WrapperPage.js [hunk 2]
```diff
@@ -30,35 +67,35 @@ const WrapperPage = ({ checker, topic }) => {
...
           heading="Adres gegevens"
           largeCircle
-        >
-          <Location />
+        >{console.log('isActive', isActive('location'))}
```
Comment (afjlambert) [inline apps/client/src/pages/WrapperPage.js]: remove `console.log`?

**Review by afjlambert — CHANGES_REQUESTED** [review]

File: apps/client/src/components/Location/Address.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import { Paragraph } from "@datapunt/asc-ui";
```
Comment (robinpiets) [inline apps/client/src/components/Location/Address.js]: I would like to rename the file `Address` to `LocationInput` and `Location` to `LocationResult`. Address is just too general. This also means renaming `setActiveState("address")` to `setActiveState("locationResult")`.

What do you think?

File: apps/client/src/components/Location/Address.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  return (
+    <Form onSubmit={onSubmit} data-testid={ADDRESS_PAGE}>
+      <RegisterLookupSummary
+        displayZoningPlans={!useSTTR}
+        address={address}
+        setActiveState={setActiveState}
+        setFinishedState={setFinishedState}
+        topic={topic}
+      />
+      <Paragraph gutterBottom={useSTTR && topic.text?.addressPage ? null : 0}>
```
Comment (robinpiets) [inline apps/client/src/components/Location/Address.js]: Something went wrong in merging with the chappie2.0 branch, because this whole text has changed:

On chappie2.0 branch:
```
<Paragraph gutterBottom={useSTTR ? null : 0}>
  {/* OLO Flow text */}
  {!useSTTR &&
    ` U hebt deze informatie nodig om de vergunningcheck te doen op
    het Omgevingsloket.`}
</Paragraph>
```

Please check carefully what text to place in this file

File: apps/client/src/components/Location/Address.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          : // OLO Flow text
+            ` U hebt deze informatie nodig om de vergunningcheck te doen op
+              het Omgevingsloket.`}
+      </Paragraph>
+
+      {/* Extra text about this activity (text that can be in both flows) */}
+      {/* This is also text we need to discuss because it's not in new design */}
+      {topic.text?.addressPage && (
+        <Paragraph gutterBottom={0}>{topic.text.addressPage}</Paragraph>
+      )}
```
Comment (robinpiets) [inline apps/client/src/components/Location/Address.js]: Something went wrong in merging with the chappie2.0 branch, because this whole text has been removed. Please check carefully what text to place in this file

File: apps/client/src/components/Location/Address.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    // Generate OLO parameter "streetNumber"
+    const oloStreetNumber = `facet_locatie_huisnummer=${houseNumber}`;
+
+    const oloSuffix = `facet_locatie_huisnummertoevoeging=${houseNumberFull
+      .replace(houseNumber, "")
+      .trim()}`;
+
+    // Redirect user to OLO with all parameters
+    return `${OLO.location}?param=postcodecheck&${oloPostalCode}&${oloStreetNumber}&${oloSuffix}`;
+  };
```
Comment (robinpiets) [inline apps/client/src/components/Location/Address.js]: Shall we move this to config/index.ts? We can do this in a separate PR, I am adding this to the ToDo list.

File: apps/client/src/components/Location/Address.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import Nav from "../Nav";
+import RegisterLookupSummary from "../RegisterLookupSummary";
+
+const Address = ({ topic, setFinishedState, setActiveState, isFinished }) => {
+  const sessionContext = useContext(SessionContext);
+  const address = sessionContext[topic.slug].address || {};
+  const useSTTR = !!topic.sttrFile;
+
+  const getOloUrl = ({ postalCode, houseNumberFull, houseNumber }) => {
+    // Form is validated, we can proceed
```
Comment (robinpiets) [inline apps/client/src/components/Location/Address.js]: This comment is in the wrong place

File: apps/client/src/components/Location/Location.js [hunk 8]
```diff
@@ -76,83 +70,10 @@ const Location = ({ topic, resetChecker }) => {
       if (focus) {
         document.activeElement.blur();
       } else {
-        sessionContext.setSessionData([
-          slug,
-          {
-            addressShown: true,
-          },
-        ]);
+        setActiveState("address");
```
Comment (robinpiets) [inline apps/client/src/components/Location/Location.js]: There is still `finishedLocation: false` in withTopic.js

File: apps/client/src/components/Questions.js [hunk 9]
```diff
@@ -33,13 +38,11 @@ const Questions = ({ checker, topic: { slug } }) => {
...
-      // Go directly to "Contact Conclusion" and skip other questions
-      sessionContext.setSessionData([
-        slug,
-        {
-          finishedQuestions: true,
-        },
-      ]);
+      // Undo the next() with previous(), because we were already at the final question
+      checker.previous();
+      // Go to "Conclusion"
```
Comment (robinpiets) [inline apps/client/src/components/Questions.js]: These three lines can be removed and can be replaced with the comment: `// Go directly to "Contact Conclusion" and skip other questions`

We don't need the checker.previous() anymore

File: apps/client/src/components/Questions.js [hunk 10]
```diff
@@ -56,14 +59,12 @@ const Questions = ({ checker, topic: { slug } }) => {
...
-            {
-              finishedQuestions: true,
-            },
-          ]);
+          setActiveState("conclusion");
+          setFinishedState(["questions", "conclusion"], true);
         }
       } else {
+        setActiveState("conclusion");
+        setFinishedState(["questions", "conclusion"], true);
```
Comment (robinpiets) [inline apps/client/src/components/Questions.js]: This should not enable the conclusion right? Why should this be here?

File: apps/client/src/components/Questions.js [hunk 11]
```diff
@@ -85,29 +86,23 @@ const Questions = ({ checker, topic: { slug } }) => {
...
-      ]);
     }
   };
 
   const onGoToQuestion = (questionId) => {
     // Checker rewinding also needs to work when you already have a conlusion
     // Go to the specific question in the stack
+    checker.rewindTo(questionId);
+    setActiveState("questions");
+    setFinishedState(["conclusion", "questions"], false);
```
Comment (robinpiets) [inline apps/client/src/components/Questions.js:93]: Shall we wrap this in an `if (isFinished("conclusion"))...`?

File: apps/client/src/pages/IntroPage.js [hunk 12]
```diff
@@ -1,28 +1,19 @@
...
   const Intro = React.lazy(() => import(`../intros/${intro}`));
 
   const goToNext = () => {
-    sessionContext.setSessionData([
-      slug,
-      {
-        finishedLocation: false,
-        finishedQuestions: false,
-      },
-    ]);
```
Comment (robinpiets) [inline apps/client/src/pages/IntroPage.js:25]: Do we not need some sort of reset here?

File: apps/client/src/pages/WrapperPage.js [hunk 13]
```diff
@@ -17,14 +18,45 @@ import withChecker from "../hoc/withChecker";
 const WrapperPage = ({ checker, topic, resetChecker }) => {
   const sessionContext = useContext(SessionContext);
   const { slug, sttrFile } = topic;
-  const { finishedLocation, finishedQuestions } = sessionContext[slug] || false;
+  const { activeComponents, finishedComponents } = sessionContext[slug] || []; // At startup we don't have activeComponent or finishedComponents, so start with empty array.
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: Can you place the comment above?

File: apps/client/src/pages/WrapperPage.js [hunk 14]
```diff
@@ -17,14 +18,45 @@ import withChecker from "../hoc/withChecker";
...
+
+  // Only one component can be active at the same time.
+  const setActiveState = (component) => {
+    sessionContext.setSessionData([slug, { activeComponents: [component] }]);
+  };
+
+  // Pass a component name, or a array of components, this array or component will all be finished or removed.
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: What do you mean with `will all be finished or removed.`?

File: apps/client/src/pages/WrapperPage.js [hunk 15]
```diff
@@ -17,14 +18,45 @@ import withChecker from "../hoc/withChecker";
...
+  const setFinishedState = (component, value) => {
+    let newFinishedComponents = finishedComponents || []; // If there are no finishedComponents start with a empty array.
+    const allComponents = Array.isArray(component) ? component : [component];
+    if (typeof value === "boolean" && value === false) {
+      newFinishedComponents =
+        finishedComponents?.filter((c) => !allComponents.includes(c)) || [];
+    } else {
+      newFinishedComponents = [...newFinishedComponents, ...allComponents];
+    }
+    // Save the new sessionData to the contex
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: contex -> context

File: apps/client/src/pages/WrapperPage.js [hunk 16]
```diff
@@ -17,14 +18,45 @@ import withChecker from "../hoc/withChecker";
...
+      slug,
+      { finishedComponents: newFinishedComponents },
+    ]);
+  };
+
+  // @TODO: maybe this code can be refactored to one function, but i think this is really declarative. Lets discuss
+  const isActive = (component) => {
+    const allComponents = Array.isArray(component) ? component : [component];
+    return activeComponents?.includes(...allComponents) || false;
+  };
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: If it's basically the same code than we can better use just one, or maybe something like this: ` const isFinished = (value) => isActive(value)`

File: apps/client/src/pages/WrapperPage.js [hunk 17]
```diff
@@ -35,36 +67,74 @@ const WrapperPage = ({ checker, topic, resetChecker }) => {
             highlightActive
           >
             <StepByStepItem
-              active={!finishedLocation}
-              checked={finishedLocation}
+              active={
+                (isActive("location") ||
+                  isActive("address") ||
+                  !activeComponents) &&
+                !isFinished("address")
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: Holy moly, this is a huge condition. Can we not just use `!isFinished("address")`?

File: apps/client/src/pages/WrapperPage.js [hunk 18]
```diff
@@ -35,36 +67,74 @@ const WrapperPage = ({ checker, topic, resetChecker }) => {
...
+              }
+              checked={isFinished("address")}
               heading="Adres gegevens"
               largeCircle
             >
-              <Location resetChecker={resetChecker} checker={checker} />
+              {(!activeComponents || isActive("location")) && (
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: Can we not set an active component on first init `if (!activeComponents) ...`?

File: apps/client/src/pages/WrapperPage.js [hunk 19]
```diff
@@ -35,36 +67,74 @@ const WrapperPage = ({ checker, topic, resetChecker }) => {
...
+                  resetChecker={resetChecker}
+                  setActiveState={setActiveState}
+                />
+              )}
+              {(isActive("address") || isFinished("address")) && (
+                <Address
+                  topic={topic}
+                  isFinished={isFinished}
+                  setActiveState={setActiveState}
+                  setFinishedState={setFinishedState}
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: you can shorten this to 
```
{...{
topic, 
isFinished, 
setActiveState,
setFinishedState,
}}
```

File: apps/client/src/pages/WrapperPage.js [hunk 20]
```diff
@@ -35,36 +67,74 @@ const WrapperPage = ({ checker, topic, resetChecker }) => {
...
+              checked={isFinished("questions")}
+              customSize
+              done={isFinished("address")}
               heading="Vragen"
               largeCircle
             />
-            {finishedLocation && <Questions checker={checker} topic={topic} />}
+            {isFinished("address") && (
+              <Questions
+                checker={checker}
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: you can shorten this

File: apps/client/src/pages/WrapperPage.js [hunk 21]
```diff
@@ -35,36 +67,74 @@ const WrapperPage = ({ checker, topic, resetChecker }) => {
...
+          <Location topic={topic} setActiveState={setActiveState} />
+        )}
+        {!sttrFile && isActive("address") && (
+          <Address
+            topic={topic}
+            isFinished={isFinished}
+            setActiveState={setActiveState}
+            setFinishedState={setFinishedState}
+          />
+        )}
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: This part can be shortened

**Review by robinpiets — CHANGES_REQUESTED** [review]

File: apps/client/src/pages/WrapperPage.js (hunk 14 above)
Comment (robinpiets) [reply apps/client/src/pages/WrapperPage.js]: Please enhance the second parameter

File: apps/client/src/pages/IntroPage.js (hunk 12 above)
Comment (robinpiets) [reply apps/client/src/pages/IntroPage.js:25]: something like `setActive("locationInput")` here?

File: apps/client/src/components/Location/LocationResult.js [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    if (useSTTR) {
+      setFinishedState("locationResult", true);
+      setActiveState("questions");
+    } else {
+      window.open(getOloUrl(address), "_blank");
+    }
+  };
+
+  return (
+    <Form onSubmit={onSubmit} data-testid={ADDRESS_PAGE}>
```
Comment (robinpiets) [inline apps/client/src/components/Location/LocationResult.js]: Can you rename `ADDRESS_PAGE` to `LOCATION_RESULT`?

File: apps/client/src/components/Location/LocationResult.js [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+      <RegisterLookupSummary
+        displayZoningPlans={!useSTTR}
+        address={address}
+        setActiveState={setActiveState}
+        setFinishedState={setFinishedState}
+        topic={topic}
+      />
+      <Paragraph gutterBottom={useSTTR ? null : 0}>
+        {/* OLO Flow text */}
+        {!useSTTR &&
```
Comment (robinpiets) [inline apps/client/src/components/Location/LocationResult.js]: As discussed...

File: apps/client/src/pages/WrapperPage.js [hunk 24]
```diff
@@ -17,14 +18,50 @@ import withChecker from "../hoc/withChecker";
 const WrapperPage = ({ checker, topic, resetChecker }) => {
   const sessionContext = useContext(SessionContext);
   const { slug, sttrFile } = topic;
-  const { finishedLocation, finishedQuestions } = sessionContext[slug] || false;
+
+  // At startup we don't have activeComponent or finishedComponents, so start with empty array.
+  const { activeComponents, finishedComponents } = sessionContext[slug];
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: The end of the comment `so start with empty array.` doesn't make sense anymore?

File: apps/client/src/pages/WrapperPage.js [hunk 25]
```diff
@@ -17,14 +18,50 @@ import withChecker from "../hoc/withChecker";
...
+
+  // Only one component can be active at the same time.
+  const setActiveState = (component) => {
+    sessionContext.setSessionData([slug, { activeComponents: [component] }]);
+  };
+
+  /**
+   * Set given components to the given finished state
+   *
+   * @param { Object[] | string } component - name of names of the components
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: Nice!

Only it's `name or names` instead of `name of names`

File: apps/client/src/pages/WrapperPage.js [hunk 26]
```diff
@@ -17,14 +18,50 @@ import withChecker from "../hoc/withChecker";
...
+   * @param { boolean } value - Set components to finished state or remove from finished state.
+   */
+  const setFinishedState = (component, value) => {
+    let newFinishedComponents = finishedComponents; // If there are no finishedComponents start with a empty array.
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: Comment makes no sense anymore: `// If there are no finishedComponents start with a empty array.`

File: apps/client/src/pages/WrapperPage.js [hunk 27]
```diff
@@ -17,14 +18,50 @@ import withChecker from "../hoc/withChecker";
...
+        finishedComponents?.filter((c) => !allComponents.includes(c)) || [];
+    } else {
+      newFinishedComponents = [...newFinishedComponents, ...allComponents];
+    }
+    // Save the new sessionData to the context
+    sessionContext.setSessionData([
+      slug,
+      { finishedComponents: newFinishedComponents },
+    ]);
+  };
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: You can refactor this whole part to something like this:

```
// Comment
const allComponents = Array.isArray(component) ? component : [component];

// Comment
const newFinishedComponents =
  typeof value === "boolean" && value === false
    ? finishedComponents?.filter((c) => !allComponents.includes(c)) || []
    : [...finishedComponents, ...allComponents];

// Comment
sessionContext.setSessionData([slug, ...newFinishedComponents]);
```

File: apps/client/src/pages/WrapperPage.js [hunk 28]
```diff
@@ -17,14 +18,50 @@ import withChecker from "../hoc/withChecker";
...
+
+  const isActive = (component, finished = false) => {
+    const components = finished ? finishedComponents : activeComponents;
+    const allComponents = Array.isArray(component) ? component : [component];
+    return components.includes(...allComponents) || false;
+  };
+
+  const isFinished = (component) => isActive(component, true);
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: Cool!

File: apps/client/src/pages/WrapperPage.js [hunk 29]
```diff
@@ -35,36 +72,82 @@ const WrapperPage = ({ checker, topic, resetChecker }) => {
...
             <StepByStepItem
-              active={!finishedLocation}
-              checked={finishedLocation}
+              active={isActive("locationInput") || isActive("locationResult")}
+              checked={isFinished("locationResult")}
               heading="Adres gegevens"
               largeCircle
             >
-              <Location resetChecker={resetChecker} checker={checker} />
+              {(!activeComponents || isActive("locationInput")) && (
```
Comment (robinpiets) [inline apps/client/src/pages/WrapperPage.js]: You didn't find a way to init an activeComponent so we can drop the `!activeComponents` ?

**Review by robinpiets — CHANGES_REQUESTED** [review]
It's becoming a very nice refactor!! I've got some minor points though

**Review by robinpiets — APPROVED** [review]
